### ISSUE TITLE
"imagePullPolicy: IfNotPresent" behavior with SNAPSHOT versions

### ISSUE THREAD

**Issue by TimMoore** [lightbend/openshift-deployment-guide#19]
The combination of `imagePullPolicy: IfNotPresent` and use of SNAPSHOT versions in image tags can be confusing.

On Minishift, it works fine, because it runs a single-node OpenShift cluster. When you push the image (using `sbt docker:publishLocal` for example), it's pushing the image to the same location that it launches from. It overwrites the existing image, and all is well.

When using a production cluster, if I understand correctly, changes to the image won't always be recognized when redeploying. My (untested) understanding is that when you push to the OpenShift image registry, it will only put the new image layers into the local node that you push to, and then other nodes won't pull from that if they already have the previous image version cached.

There are a number of ways this could be solved. It's a good idea to include the git hash in the version. The guide suggests this, but doesn't show how to do it. For development, you could use the `latest` image tag instead of the `1.0-SNAPSHOT` tag. Then, you can leave out the explicit `imagePullPolicy` configuration and rely on the default, which has a special case for `latest`:

If a container’s `imagePullPolicy` parameter is not specified, OKD sets it based on the image’s tag:
1. If the tag is `latest`, OKD defaults `imagePullPolicy` to `Always`.
2. Otherwise, OKD defaults `imagePullPolicy` to `IfNotPresent`.

https://docs.okd.io/latest/dev_guide/managing_images.html#image-pull-policy

Same for commercial OpenShift and standard Kubernetes:
* https://docs.openshift.com/container-platform/3.11/dev_guide/managing_images.html#image-pull-policy
* https://kubernetes.io/docs/concepts/configuration/overview/#container-images

**Comment by TimMoore** [comment]
Alternatively, it might be a good idea to recommend against using `latest` (most Kubernetes guides recommend against it) and putting the git hashed version directly into the deployment config. That means you need to update it whenever you want to test a change, but otherwise the workflow for testing changes is not that clear. Kubernetes only performs rolling deployments when the `Deployment` itself changes, and the usual workflow is that you change the image tag to deploy a new version.

The alternative is to kill pods explicitly and let it restart them. It will then use the `imagePullPolicy` to determine what to do, which is where I think `IfNotPresent` could conflict with the use of a constant image label.

**Comment by jroper** [comment]
Another thing to consider is using deployment configs as discussed in #4. I believe one of the primary things that deployment configs offer over deployments is the ability to trigger a redeploy when you push a new image to an image stream. Of course, this is OpenShift specific.

**Comment by TimMoore** [comment]
Yes, it would be nice to have an additional section describing how to use OpenShift-specific features, but I think it's more important to show how to do things with standard Kubernetes features.

**Comment by jroper** [comment]
The challenge of course with including the git hash in the version number - from the point of view of the guide - is the need to coordinate the version number between the spec and the build. I guess our instructions to push docker images to the OpenShift registry are already not copy/pasteable, since they contain the registry host name. But it will make the guide a little harder to follow.

**Comment by TimMoore** [comment]
That might be a reason to use `latest` in the guide.

**Comment by jroper** [comment]
With `imagePullPolicy`: `Always`?

**Comment by jroper** [comment]
Maybe we should just leave `imagePullPolicy` out given the defaults.

**Comment by TimMoore** [comment]
Yes, I think we should leave it out and rely on the default, use `latest` in the examples, and revise the bullet point about the SNAPSHOT versions to instead say that for production deployments, a stable image label should be used instead of `latest` and updated to a new label to deploy a new version.

**Comment by jroper** [comment]
Ok, so here's the problem with using `latest` and no `imagePullPolicy`. It doesn't work in Minishift. After pushing to the OpenShift docker registry, if `imagePullPolicy` is `Always`, it will try to pull the image from `docker.io`, and fail. It works fine in production because the image is tagged with the OpenShift repo location.

One option might be to use Minishift in a way that's more similar to production. I'm not exactly sure what that would look like, it might mean an explicit push to the Minishift docker registry, which would mean configuring it to be this:

```
$ minishift openshift registry
172.30.1.1:5000
```

I think that IP address changes on each start of Minishift. It's a bit of a drag on the usability of the guide. We could also override `imagePullPolicy` to `IfNotPresent`.

**Comment by jroper** [comment]
Ok, I think I have a better idea of what's going on now.

Docker hosts, and docker registries, are different things.

A docker host is something that runs on any machine that wants to run docker images (also if it wants to build them). You usually only interact with it locally. You pull images into it from a registry, and you can push images from it to a registry. You can run images in it.

A docker registry is a registry that docker hosts can pull from/push to. Before a docker host can run a docker image, it first needs to pull it from a registry (if it's not already already pulled).

Minishift runs its own VM which runs a docker host (it has to, if the VM wants to run containers itself). When you run `minishift docker-env`, this gives you the environment variables to make your local docker command talk to the docker host running on the Minishift VM, rather than your own systems local docker host. This allows you to directly build docker images in it, and run/stop containers directly on it using your docker command. This docker host is what all docker containers running in the Minishift VM use, so your Kubernetes pods are created using these. That's why when you use this host, and build an image directly in it, there's no need to push anything, and pulling of course won't work if you haven't pushed the image anywhere.

When you run `minishift openshift registry`, you get the in built in OpenShift docker registry. You can't build images directly in that, you could push to it though, and then the Minishift docker host could pull from it.

So I'll have to update the docs to describe this.

**Comment by jroper** [comment]
I think I'll make the Minishift approach the same as the production approach - it's just confusing otherwise (the one way for Minishift another way for proper OpenShift environment has irked me).

I've already modified the build so that it uses the build tool (`sbt-native-packager` or Fabric8) to do the pushing. What I might do is have both default the registry to a system property called `docker.registry`. Then they can run with `-Ddocker.registry=$(minishift openshift registry)` for Minishift, that's easy, it's copy/pastable, and can also be used in the proper OpenShift docs so you don't need to modify the build files in anyway to deploy.

**Comment by srikanthops** [comment]
HI I was trying to get shoppingcartapp work in azure cluster.. 
the only change I had to do was change ImagePullPolicy  to always and it magically worked.. 

Before this it couldn't even form the cluster 
https://github.com/lightbend/lb-demos/pull/10

**Comment by TimMoore** [comment]
@srikanthops I think the best thing to do is leave it out and rely on the default behavior
